clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashMultiSet.java

```diff
@@ -0,0 +1,142 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * A memory-efficient hash multiset which tracks the order of insertion of elements.
+ * See org.apache.kafka.common.utils.ImplicitLinkedHashSet for implementation details.
+ *
+ * This class is a multi-set because it allows multiple elements to be inserted that are
+ * equal to each other.
+ *
+ * We use reference equality when adding elements to the set.  A new element A can
+ * be added if there is no existing element B such that A == B.  If an element B
+ * exists such that A.equals(B), A will still be added.
+ *
+ * When deleting an element A from the set, we will try to delete the element B such
+ * that A == B.  If no such element can be found, we will try to delete an element B
+ * such that A.equals(B).
+ *
+ * contains() and find() are unchanged from the base class-- they will look for element
+ * based on object equality, not reference equality.
+ *
+ * This multiset does not allow null elements.  It does not have internal synchronization.
+ */
+public class ImplicitLinkedHashMultiSet<E extends ImplicitLinkedHashSet.Element>
+        extends ImplicitLinkedHashSet<E> {
+    public ImplicitLinkedHashMultiSet() {
+        super(0);
+    }
+
+    public ImplicitLinkedHashMultiSet(int expectedNumElements) {
+        super(expectedNumElements);
+    }
+
+    public ImplicitLinkedHashMultiSet(Iterator<E> iter) {
+        super(iter);
+    }
+
+
+    /**
+     * Adds a new element to the appropriate place in the elements array.
+     *
+     * @param newElement    The new element to add.
+     * @param addElements   The elements array.
+     * @return              The index at which the element was inserted, or INVALID_INDEX
+     *                      if the element could not be inserted.
+     */
+    @Override
+    int addInternal(Element newElement, Element[] addElements) {
+        int slot = slot(addElements, newElement);
+        for (int seen = 0; seen < addElements.length; seen++) {
+            Element element = addElements[slot];
+            if (element == null) {
+                addElements[slot] = newElement;
+                return slot;
+            }
+            if (element == newElement) {
+                return INVALID_INDEX;
+            }
+            slot = (slot + 1) % addElements.length;
+        }
+        throw new RuntimeException("Not enough hash table slots to add a new element.");
+    }
+
+    /**
+     * Find an element matching an example element.
+     *
+     * @param key               The element to match.
+     *
+     * @return                  The match index, or INVALID_INDEX if no match was found.
+     */
+    @Override
+    int findElementToRemove(Object key) {
+        if (key == null) {
+            return INVALID_INDEX;
+        }
+        int slot = slot(elements, key);
+        int bestSlot = INVALID_INDEX;
+        for (int seen = 0; seen < elements.length; seen++) {
```

Comment:
I'm not sure `foreach` is really better here, since the variable that we use would be equally unused.  This loop itself is kind of a bit of paranoia on my part.  It prevents an infinite loop if the hash table is 100% full.  But the hash table should never get more than 50% full anyway...